mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -45,19 +47,53 @@ impl S3CrtClient {
                 .set_header(&Header::new("x-amz-storage-class", storage_class))
                 .map_err(S3RequestError::construction_failure)?;
         }
-
+        let mut maybe_sse = None;
+        let mut maybe_key_id = None;
+        match params.server_side_encryption.to_owned() {
+            ServerSideEncryption::Default => (),
+            ServerSideEncryption::Kms { key_id } => {
+                maybe_sse = Some("aws:kms");
+                maybe_key_id = key_id;
+            }
+            ServerSideEncryption::DualLayerKms { key_id } => {
+                maybe_sse = Some("aws:kms:dsse");
+                maybe_key_id = key_id;
+            }
+        }
+        let mut expected_headers = Vec::new();
+        if let Some(sse) = maybe_sse {
+            let header_name = "x-amz-server-side-encryption";
+            expected_headers.push((header_name.to_owned(), sse.to_owned()));
+            message
+                .set_header(&Header::new(header_name, sse))
+                .map_err(S3RequestError::construction_failure)?;
+        }
+        if let Some(key_id) = maybe_key_id {
+            let header_name = "x-amz-server-side-encryption-aws-kms-key-id";
+            expected_headers.push((header_name.to_owned(), key_id.clone()));
+            message
+                .set_header(&Header::new(header_name, key_id))
+                .map_err(S3RequestError::construction_failure)?;
+        }
+        let response_headers: Arc<Mutex<Option<Headers>>> = Default::default();
```

Comment:
`response_headers` will be accessed from different threads -- from CRT thread which executes `on_headers` callback and from our thread which executes `review_and_complete`, practically there is no need for synchronisation here as it is known that those accesses will be always sequential (`review_and_complete` is guaranteed to be called when `on_headers` has already completed); nevertheless compiler is not smart enough to prove that, so we need a mutex here to avoid using `unsafe`

Reply:
This explanation should be in comments in the code.

Reply:
Thanks for adding it, it is good to have a documented reason rather than just guessing or not knowing.